core/src/types/operator/registry.rs

```diff
@@ -0,0 +1,242 @@
+    pub fn new() -> Self {
+        Self {
+            registry: HashMap::new(),
+        }
+    }
+
+    pub fn register(&mut self, scheme: &str, factory: OperatorFactory) {
+        // TODO: should we receive a `&str` or a `String`? we are cloning it anyway
+        self.registry.insert(scheme.to_string(), factory);
+    }
+
+    pub fn parse(
+        &self,
+        uri: &str,
+        options: impl IntoIterator<Item = (String, String)>,
+    ) -> Result<Operator> {
+        // TODO: we use the `url::Url` struct instead of `http:Uri`, because
+        // we needed it in `Configurator::from_uri` method.
+        let parsed_url = url::Url::parse(uri).map_err(|err| {
+            Error::new(ErrorKind::ConfigInvalid, "uri is invalid")
+                .with_context("uri", uri)
+                .set_source(err)
+        })?;
+
+        // TODO: with the `url::Url` struct, we always have the scheme (it is not an Option<str>)
+        // but with the `http::Uri` crate, it can be missing https://docs.rs/http/latest/http/uri/struct.Uri.html#method.scheme
+        // which one should we use?
+        let scheme = parsed_url.scheme();
+
+        let factory = self.registry.get(scheme).ok_or_else(|| {
+            Error::new(
+                ErrorKind::ConfigInvalid,
+                "could not find any operator factory for the given scheme",
+            )
+            .with_context("uri", uri)
+            .with_context("scheme", scheme)
+        })?;
+
+        // TODO: `OperatorFactory` should receive `IntoIterator<Item = (String, String)>` instead of `HashMap<String, String>`?
+        // however, impl Traits in type aliases is unstable and also are not allowed in fn pointers
+        let options = options.into_iter().collect();
+
+        // TODO: `OperatorFactory` should use `&str` instead of `String`? we are cloning it anyway
+        factory(uri, options)
+    }
+
+    pub fn with_enabled_services() -> Self {
+        let mut registry = Self::new();
+        // TODO: is this correct? have a `Builder::enabled()` method that returns the set of enabled services builders?
+        // Similar to `Scheme::Enabled()`
+        // or have an `Scheme::associated_builder` that given a scheme returns the associated builder? The problem with this
+        // is that `Scheme` variants are not gate behind a feature gate and the associated builder is. As a workaround
+
+        // TODO: it seems too error-prone to have this list manually updated, we should have a macro that generates this list?
+        // it could be something like:
+        //
+        // ```rust
+        // apply_for_all_services!{
+        //    registry.register_builder::<$service>();
+        // }
+        // ```
+        // and the apply_for_all_services macro would gate every statement behind the corresponding feature gate
+        // This seems to not be the place where we should have a "list of enabled services".
+        #[cfg(feature = "services-aliyun-drive")]
```

Comment:
I believe every service should have its own registration functions. There are two reasons:

- The service may have different schemes for registration. For example, S3 registers as `s3`, `s3a`, `minio`, `r2`, and so on, while GCS registers as `gcs`, `gs`, and so forth.
- Provide a register that simplifies integration with OpenDAL for external users.